python/ql/consistency-queries/DataFlowConsistency.ql

```diff
@@ -0,0 +1,122 @@
+    // argument in the CallNode, we end up violating this consistency check in those
+    // cases. (see `getCallArg` in DataFlowDispatch.qll)
+    exists(DataFlowCall other, CallNode cfgCall | other != call |
+      call.getNode() = cfgCall and
+      other.getNode() = cfgCall and
+      isArgumentNode(arg, call, _) and
+      isArgumentNode(arg, other, _)
+    )
+    or
+    // bound methods that refer to the same self argument.
+    // Example: In `bm = self.foo; bm(); bm()` both bm() calls use the same `self` as
+    // the (pos self) argument
+    exists(AttrRead attr, DataFlowCall other | other != call |
+      // for simple cases we can track the function back to the attr read but when the
+      // call appears in the body of a list-comprehension, we can't do that, and simply
+      // allow it instead.
+      (
+        call.getScope() = attr.getScope() and
+        any(CfgNode n | n.asCfgNode() = call.getNode().(CallNode).getFunction()).getALocalSource() =
+          attr
+        or
+        not exists(call.getScope().(Function).getDefinition()) and
+        call.getScope().getScope+() = attr.getScope()
+      ) and
+      (
+        other.getScope() = attr.getScope() and
+        any(CfgNode n | n.asCfgNode() = other.getNode().(CallNode).getFunction()).getALocalSource() =
+          attr
+        or
+        not exists(other.getScope().(Function).getDefinition()) and
+        other.getScope().getScope+() = attr.getScope()
+      ) and
+      arg = attr.getObject() and
+      arg = call.getArgument(any(ArgumentPosition p | p.isSelf())) and
+      arg = other.getArgument(any(ArgumentPosition p | p.isSelf()))
+    )
+    or
+    // `f = getattr(obj, "foo"); f()` where `obj` is used as (pos self) argument for
+    // `f()` call
+    exists(DataFlowCall getAttrCall, DataFlowCall methodCall, AttrRead attr |
+      call in [getAttrCall, methodCall]
+    |
+      arg = getAttrCall.getArgument(any(ArgumentPosition p | p.isPositional(0))) and
+      arg = methodCall.getArgument(any(ArgumentPosition p | p.isSelf())) and
+      attr.getObject() = arg and
+      attr.(CfgNode).getNode() = getAttrCall.getNode()
+    )
+    or
+    // In the code `super(Base, self).foo()` we use `self` as an argument in both the
+    // super() call (pos 1) and in the .foo() call (pos self).
+    exists(DataFlowCall superCall, DataFlowCall methodCall | call in [superCall, methodCall] |
+      exists(superCallTwoArgumentTracker(_, arg)) and
+      arg = superCall.getArgument(any(ArgumentPosition p | p.isPositional(1))) and
+      arg = methodCall.getArgument(any(ArgumentPosition p | p.isSelf()))
+    )
+    or
+    // in the code `def func(self): super().foo(); super.bar()` we use `self` as the
+    // (pos self) argument in both .foo() and .bar() calls.
+    exists(Function f |
+      exprNode(f.getArg(0)) = arg and
+      call.getNode().getScope() = f and
+      arg = call.getArgument(any(ArgumentPosition p | p.isSelf()))
+    )
```

Comment:
I would think you want an `other` in this logic as well? Otherwise you may be hiding other problems?

Reply:
good point, done in https://github.com/github/codeql/pull/8457/commits/02f20312395a757435ac9710183fbcfc613ebf07